File: ABC_Classic/ABC_classic.py
# abc_classic.py
import random
import time
import math
from datetime import timedelta
from typing import List, Callable, Optional, Tuple, Union
import numpy as np
import matplotlib.pyplot as plt


class ABC_Classic:
    """
    Классическая реализация алгоритма искусственной пчелиной колонии (ABC)
    для задачи коммивояжёра (TSP).
    
    Поддерживает инициализацию через fitness_function или distance_matrix.
    """
    
    # ========================================================================
    # Внутренний класс: базовая пчела
    # ========================================================================
    class _Bee:
        def __init__(self, solution: List[int], fitness_func: Callable):
            self.solution = solution.copy()
            self.fitness_func = fitness_func
            self.fitness = fitness_func(solution)
            self.trial = 0

        def _evaluate(self, sol: List[int]) -> float:
            return self.fitness_func(sol)

        def _greedy_update(self, new_sol: List[int]) -> bool:
            new_fit = self._evaluate(new_sol)
            if new_fit > self.fitness:
                self.solution = new_sol
                self.fitness = new_fit
                self.trial = 0
                return True
            self.trial += 1
            return False

    # ========================================================================
    # Внутренний класс: рабочая пчела (Employed Bee)
    # ========================================================================
    class _EmployedBee(_Bee):
        def explore(self, population: List['_Bee']) -> bool:
            candidates = [b for b in population if b is not self]
            if not candidates:
                return False
            partner = random.choice(candidates)
            new_sol = self._crossover_with_partner(partner.solution)
            return self._greedy_update(new_sol)

        def _crossover_with_partner(self, partner_sol: List[int]) -> List[int]:
            """OX-like кроссовер для перестановок"""
            size = len(self.solution)
            new_sol = [-1] * size
            
            start = random.randint(0, size - 2)
            end = random.randint(start + 1, min(start + size // 2, size))
            new_sol[start:end] = self.solution[start:end]
            
            ptr = 0
            for i in range(size):
                if new_sol[i] == -1:
                    while partner_sol[ptr] in new_sol:
                        ptr += 1
                    new_sol[i] = partner_sol[ptr]
                    ptr += 1
            return new_sol

    # ========================================================================
    # Внутренний класс: пчела-наблюдатель (Onlooker Bee)
    # ========================================================================
    class _OnlookerBee(_Bee):
        def explore(self, population: List['_Bee']) -> bool:
            if not population:
                return False
            probs = self._calc_selection_probs(population)
            candidates = random.choices(population, weights=probs, k=3)
            base_bee = max(candidates, key=lambda b: b.fitness)
            new_sol = self._mutate(base_bee.solution)
            return self._greedy_update(new_sol)

        def _calc_selection_probs(self, population: List['_Bee']) -> List[float]:
            fitnesses = [b.fitness for b in population]
            if not fitnesses:
                return [1/len(population)] * len(population)
            max_f = max(fitnesses)
            exp_vals = [math.exp((f - max_f) * 10) for f in fitnesses]
            total = sum(exp_vals)
            return [e/total for e in exp_vals] if total > 0 else [1/len(population)] * len(population)

        def _mutate(self, base_sol: List[int]) -> List[int]:
            new_sol = base_sol.copy()
            n = len(new_sol)
            if n < 2:
                return new_sol
            mutation = random.choice(['inversion', 'swap', 'shift'])
            if mutation == 'inversion':
                i, j = sorted(random.sample(range(n), 2))
                new_sol[i:j+1] = reversed(new_sol[i:j+1])
            elif mutation == 'swap':
                i, j = random.sample(range(n), 2)
                new_sol[i], new_sol[j] = new_sol[j], new_sol[i]
            else:
                idx = random.randint(0, n-1)
                city = new_sol.pop(idx)
                new_pos = random.randint(0, len(new_sol))
                new_sol.insert(new_pos, city)
            return new_sol

    # ========================================================================
    # Внутренний класс: пчела-разведчик (Scout Bee)
    # ========================================================================
    class _ScoutBee(_Bee):
        def __init__(self, solution: List[int], fitness_func: Callable, lb: int, ub: int):
            super().__init__(solution, fitness_func)
            self.lb = lb
            self.ub = ub

        def explore(self, _) -> bool:
            new_sol = self._generate_random()
            return self._greedy_update(new_sol)

        def _generate_random(self) -> List[int]:
            return random.sample(range(self.lb, self.ub + 1), self.ub - self.lb + 1)

    # ========================================================================
    # Основной класс ABC_Classic
    # ========================================================================
    
    def __init__(
        self,
        fitness_function: Optional[Callable[[List[int]], float]] = None,
        distance_matrix: Optional[np.ndarray] = None,
        lb: Optional[int] = None,
        ub: Optional[int] = None,
        num_employed_bees: int = 50,
        num_onlooker_bees: int = 50,
        limit: int = 20,
        patience: int = 100,
        seed: Optional[int] = None
    ):
        """
        Инициализация алгоритма.
        
        :param fitness_function: функция оценки (чем больше — тем лучше)
        :param distance_matrix: матрица расстояний (альтернатива fitness_function)
        :param lb/ub: границы нумерации городов (авто-определяются из distance_matrix)
        """
        if seed is not None:
            random.seed(seed)
            np.random.seed(seed)
            
        # Авто-генерация фитнес-функции из distance_matrix
        if distance_matrix is not None:
            self.distance_matrix = np.array(distance_matrix)
            n = len(self.distance_matrix)
            lb = lb if lb is not None else 0
            ub = ub if ub is not None else n - 1
            
            def _tsp_fitness(route: List[int]) -> float:
                total = sum(self.distance_matrix[route[i]][route[i+1]] for i in range(len(route)-1))
                total += self.distance_matrix[route[-1]][route[0]]
                return 1.0 / total if total > 0 else float('inf')
            
            self.fitness_function = _tsp_fitness
            self._distance_eval = lambda route: sum(self.distance_matrix[route[i]][route[i+1]] for i in range(len(route)-1)) + self.distance_matrix[route[-1]][route[0]]
        elif fitness_function is not None:
            self.fitness_function = fitness_function
            self._distance_eval = None  # пользователь сам знает, как считать расстояние
            self.distance_matrix = None
        else:
            raise ValueError("Требуется либо fitness_function, либо distance_matrix")
            
        self.lb = lb
        self.ub = ub
        self.n_cities = ub - lb + 1 if lb is not None and ub is not None else 0
        
        self.num_employed = num_employed_bees
        self.num_onlooker = num_onlooker_bees
        self.limit = limit
        self.patience = patience
        
        self.employed_bees: List[self._EmployedBee] = []
        self.best_solution: Optional[List[int]] = None
        self.best_fitness = float('-inf')
        
        self._convergence_history: List[float] = []
        self.wait = 0
        self.best_iteration = 0
        self.start_time = None

    def _generate_random_tour(self) -> List[int]:
        return random.sample(range(self.lb, self.ub + 1), self.n_cities)

    def _evaluate_distance(self, route: List[int]) -> float:
        """Вычисляет длину маршрута (если известна distance_matrix)"""
        if self._distance_eval:
            return self._distance_eval(route)
        # fallback: если нет матрицы, возвращаем 1/fitness
        return 1.0 / self.fitness_function(route) if self.fitness_function(route) > 0 else float('inf')

    def _initialize_population(self):
        for _ in range(self.num_employed):
            sol = self._generate_random_tour()
            bee = self._EmployedBee(sol, self.fitness_function)
            self.employed_bees.append(bee)
            if bee.fitness > self.best_fitness:
                self.best_fitness = bee.fitness
                self.best_solution = bee.solution.copy()

    def _employed_bee_phase(self):
        for bee in self.employed_bees:
            bee.explore(self.employed_bees)
            if bee.fitness > self.best_fitness:
                self.best_fitness = bee.fitness
                self.best_solution = bee.solution.copy()

    def _onlooker_bee_phase(self):
        for _ in range(self.num_onlooker):
            base_sol = random.choice(self.employed_bees).solution
            bee = self._OnlookerBee(base_sol, self.fitness_function)
            bee.explore(self.employed_bees)
            if bee.fitness > self.best_fitness:
                self.best_fitness = bee.fitness
                self.best_solution = bee.solution.copy()

    def _scout_bee_phase(self):
        for i, bee in enumerate(self.employed_bees):
            if bee.trial > self.limit:
                new_sol = self._generate_random_tour()
                new_bee = self._EmployedBee(new_sol, self.fitness_function)
                self.employed_bees[i] = new_bee
                if new_bee.fitness > self.best_fitness:
                    self.best_fitness = new_bee.fitness
                    self.best_solution = new_bee.solution.copy()

    def run(
        self,
        max_iterations: int,
        verbose: bool = True,
        log_interval: int = 10
    ) -> Tuple[List[int], float]:
        """
        Запуск алгоритма.
        
        :return: (лучший маршрут, длина маршрута)  # ← возвращаем расстояние, а не фитнес!
        """
        self._initialize_population()
        self.start_time = time.time()
        self._convergence_history = []
        
        for iteration in range(max_iterations):
            old_best = self.best_fitness
            
            self._employed_bee_phase()
            self._onlooker_bee_phase()
            self._scout_bee_phase()
            
            # Сохраняем длину маршрута для истории
            current_distance = self._evaluate_distance(self.best_solution) if self.best_solution else float('inf')
            self._convergence_history.append(current_distance)
            
            if self.best_fitness > old_best:
                self.wait = 0
                self.best_iteration = iteration
            else:
                self.wait += 1
                
            if self.wait >= self.patience:
                if verbose:
                    print(f"\n⚡ Ранняя остановка на итерации {iteration}")
                break
                
            if verbose and iteration % log_interval == 0:
                elapsed = str(timedelta(seconds=time.time() - self.start_time)).split('.')[0]
                print(f"Итерация {iteration:4d} | Время: {elapsed} | Длина: {current_distance:.2f}")
        
        best_distance = self._evaluate_distance(self.best_solution) if self.best_solution else float('inf')
        return self.best_solution, best_distance

    def get_convergence_history(self) -> List[float]:
        """Возвращает историю длин маршрутов по итерациям"""
        return self._convergence_history.copy()

    def plot_convergence(self, optimal_value: Optional[float] = None, title: str = "Сходимость ABC"):
        if not self._convergence_history:
            print("⚠ История пуста — сначала запустите run()")
            return
        plt.figure(figsize=(10, 5))
        iterations = list(range(len(self._convergence_history)))
        if optimal_value is not None:
            diff = [abs(d - optimal_value) for d in self._convergence_history]
            plt.plot(iterations, diff, linewidth=2, label='Ошибка')
            plt.axhline(y=0, color='gray', linestyle='--', alpha=0.3)
            plt.ylabel('Абсолютная ошибка')
        else:
            plt.plot(iterations, self._convergence_history, linewidth=2)
            plt.ylabel('Длина маршрута')
        plt.xlabel('Итерация')
        plt.title(title)
        plt.grid(alpha=0.3)
        plt.legend()
        plt.tight_layout()
        plt.show()

    def get_elapsed_time(self) -> float:
        if self.start_time is None:
            return 0
        return time.time() - self.start_time

File: abc_rust/src/abc_classic.rs
//! Classical ABC for TSP (port of `ABC_Classic/ABC_classic.py`).

use crate::tsp::{fitness_from_distance, DistanceMatrix};
use rand::seq::SliceRandom;
use rand::Rng;
use std::time::{Duration, Instant};

pub type Tour = Vec<usize>;

#[derive(Clone)]
struct EmployedBee {
    solution: Tour,
    fitness: f64,
    trial: usize,
}

impl EmployedBee {
    fn new(solution: Tour, fitness: f64) -> Self {
        Self {
            solution,
            fitness,
            trial: 0,
        }
    }
}

/// Order crossover (OX), same as Python `EmployedBee._crossover_with_partner`.
fn ox_child<R: Rng + ?Sized>(rng: &mut R, a: &[usize], b: &[usize]) -> Tour {
    let size = a.len();
    if size < 2 {
        return a.to_vec();
    }
    let mut child: Vec<Option<usize>> = vec![None; size];
    let start = rng.gen_range(0..size - 1);
    let end = rng.gen_range(start + 1..=(start + size / 2).min(size));
    for i in start..end {
        child[i] = Some(a[i]);
    }
    let mut ptr = 0usize;
    for i in 0..size {
        if child[i].is_none() {
            while ptr < size {
                let v = b[ptr];
                ptr += 1;
                if !child.iter().any(|&c| c == Some(v)) {
                    child[i] = Some(v);
                    break;
                }
            }
        }
    }
    child.into_iter().map(|x| x.unwrap_or(0)).collect()
}

fn random_tour<R: Rng + ?Sized>(rng: &mut R, n: usize) -> Tour {
    let mut t: Vec<usize> = (0..n).collect();
    t.shuffle(rng);
    t
}

fn calc_selection_probs(employed: &[EmployedBee]) -> Vec<f64> {
    if employed.is_empty() {
        return Vec::new();
    }
    let max_f = employed
        .iter()
        .map(|b| b.fitness)
        .fold(f64::NEG_INFINITY, f64::max);
    let exp_vals: Vec<f64> = employed
        .iter()
        .map(|b| ((b.fitness - max_f) * 10.0).exp())
        .collect();
    let total: f64 = exp_vals.iter().sum();
    if total > 0.0 {
        exp_vals.iter().map(|e| e / total).collect()
    } else {
        vec![1.0 / employed.len() as f64; employed.len()]
    }
}

fn pick_weighted<R: Rng + ?Sized>(rng: &mut R, weights: &[f64]) -> usize {
    let total: f64 = weights.iter().sum();
    if weights.is_empty() {
        return 0;
    }
    if total <= 0.0 {
        return rng.gen_range(0..weights.len());
    }
    let r = rng.gen::<f64>() * total;
    let mut c = 0.0;
    for (i, w) in weights.iter().enumerate() {
        c += w;
        if r < c {
            return i;
        }
    }
    weights.len() - 1
}

fn mutate_tour<R: Rng + ?Sized>(rng: &mut R, base: &[usize]) -> Tour {
    let mut new_sol = base.to_vec();
    let n = new_sol.len();
    if n < 2 {
        return new_sol;
    }
    match rng.gen_range(0..3) {
        0 => {
            if n >= 2 {
                let i = rng.gen_range(0..n);
                let j = loop {
                    let x = rng.gen_range(0..n);
                    if x != i {
                        break x;
                    }
                };
                let (lo, hi) = if i < j { (i, j) } else { (j, i) };
                new_sol[lo..=hi].reverse();
            }
        }
        1 => {
            let i = rng.gen_range(0..n);
            let mut j = rng.gen_range(0..n);
            while i == j && n > 1 {
                j = rng.gen_range(0..n);
            }
            new_sol.swap(i, j);
        }
        _ => {
            let idx = rng.gen_range(0..n);
            let city = new_sol.remove(idx);
            let new_pos = rng.gen_range(0..new_sol.len().saturating_add(1));
            new_sol.insert(new_pos, city);
        }
    }
    new_sol
}

pub struct AbcClassic {
    dm: DistanceMatrix,
    num_employed: usize,
    num_onlooker: usize,
    limit: usize,
    patience: usize,
    employed_bees: Vec<EmployedBee>,
    pub best_solution: Option<Tour>,
    best_fitness: f64,
    wait: usize,
    best_iteration: usize,
    convergence_history: Vec<f64>,
    run_started_at: Option<Instant>,
    last_run_elapsed: Option<Duration>,
    last_run_iterations: Option<usize>,
}

impl AbcClassic {
    pub fn new(
        dm: DistanceMatrix,
        num_employed_bees: usize,
        num_onlooker_bees: usize,
        limit: usize,
        patience: usize,
    ) -> Self {
        Self {
            dm,
            num_employed: num_employed_bees,
            num_onlooker: num_onlooker_bees,
            limit,
            patience,
            employed_bees: Vec::new(),
            best_solution: None,
            best_fitness: f64::NEG_INFINITY,
            wait: 0,
            best_iteration: 0,
            convergence_history: Vec::new(),
            run_started_at: None,
            last_run_elapsed: None,
            last_run_iterations: None,
        }
    }

    pub fn last_run_iterations(&self) -> Option<usize> {
        self.last_run_iterations
    }

    pub fn last_run_elapsed(&self) -> Option<Duration> {
        self.last_run_elapsed
    }

    pub fn last_run_elapsed_secs(&self) -> Option<f64> {
        self.last_run_elapsed.map(|d| d.as_secs_f64())
    }

    fn tour_distance(&self, tour: &[usize]) -> f64 {
        self.dm.tour_length(tour)
    }

    fn initialize_population<R: Rng + ?Sized>(&mut self, rng: &mut R) {
        self.employed_bees.clear();
        self.best_fitness = f64::NEG_INFINITY;
        self.best_solution = None;
        let n = self.dm.n();
        for _ in 0..self.num_employed {
            let sol = random_tour(rng, n);
            let fit = fitness_from_distance(self.tour_distance(&sol));
            let bee = EmployedBee::new(sol, fit);
            if bee.fitness > self.best_fitness {
                self.best_fitness = bee.fitness;
                self.best_solution = Some(bee.solution.clone());
            }
            self.employed_bees.push(bee);
        }
    }

    fn employed_phase<R: Rng + ?Sized>(&mut self, rng: &mut R) {
        let n = self.employed_bees.len();
        for i in 0..n {
            let new_sol = {
                let bees = &self.employed_bees;
                let partners: Vec<usize> = (0..n).filter(|&j| j != i).collect();
                if partners.is_empty() {
                    continue;
                }
                let pj = *partners.choose(rng).unwrap();
                ox_child(rng, &bees[i].solution, &bees[pj].solution)
            };
            let new_fit = fitness_from_distance(self.tour_distance(&new_sol));
            let bee = &mut self.employed_bees[i];
            if new_fit > bee.fitness {
                bee.solution = new_sol;
                bee.fitness = new_fit;
                bee.trial = 0;
                if new_fit > self.best_fitness {
                    self.best_fitness = new_fit;
                    self.best_solution = Some(bee.solution.clone());
                }
            } else {
                bee.trial += 1;
            }
        }
    }

    fn onlooker_phase<R: Rng + ?Sized>(&mut self, rng: &mut R) {
        if self.employed_bees.is_empty() {
            return;
        }
        let probs = calc_selection_probs(&self.employed_bees);
        for _ in 0..self.num_onlooker {
            let init = self.employed_bees.choose(rng).unwrap();
            let mut my_fit = init.fitness;

            let mut best_ref = &self.employed_bees[0];
            for _ in 0..3 {
                let idx = pick_weighted(rng, &probs);
                let b = &self.employed_bees[idx];
                if b.fitness > best_ref.fitness {
                    best_ref = b;
                }
            }
            let new_sol = mutate_tour(rng, &best_ref.solution);
            let new_fit = fitness_from_distance(self.tour_distance(&new_sol));
            if new_fit > my_fit {
                my_fit = new_fit;
                if my_fit > self.best_fitness {
                    self.best_fitness = my_fit;
                    self.best_solution = Some(new_sol);
                }
            }
        }
    }

    fn scout_phase<R: Rng + ?Sized>(&mut self, rng: &mut R) {
        let n = self.dm.n();
        for i in 0..self.employed_bees.len() {
            if self.employed_bees[i].trial > self.limit {
                let new_sol = random_tour(rng, n);
                let nf = fitness_from_distance(self.tour_distance(&new_sol));
                self.employed_bees[i] = EmployedBee::new(new_sol, nf);
                if nf > self.best_fitness {
                    self.best_fitness = nf;
                    self.best_solution = Some(self.employed_bees[i].solution.clone());
                }
            }
        }
    }

    pub fn run<R: Rng + ?Sized>(
        &mut self,
        rng: &mut R,
        max_iterations: usize,
        verbose: bool,
        log_interval: usize,
    ) -> (Tour, f64) {
        self.initialize_population(rng);
        self.wait = 0;
        self.best_iteration = 0;
        self.convergence_history.clear();
        self.run_started_at = Some(Instant::now());
        self.last_run_elapsed = None;
        self.last_run_iterations = None;

        for iteration in 0..max_iterations {
            let old_best = self.best_fitness;
            self.employed_phase(rng);
            self.onlooker_phase(rng);
            self.scout_phase(rng);

            let current_distance = self
                .best_solution
                .as_ref()
                .map(|s| self.tour_distance(s))
                .unwrap_or(f64::INFINITY);
            self.convergence_history.push(current_distance);

            if self.best_fitness > old_best {
                self.wait = 0;
                self.best_iteration = iteration;
            } else {
                self.wait += 1;
            }

            self.last_run_iterations = Some(iteration + 1);

            if self.wait >= self.patience {
                if verbose {
                    eprintln!("Early stop at iteration {iteration}");
                }
                break;
            }

            if verbose && log_interval > 0 && iteration % log_interval == 0 {
                let elapsed = self
                    .run_started_at
                    .map(|t| t.elapsed().as_secs_f64())
                    .unwrap_or(0.0);
                eprintln!(
                    "iter {:4} | time {:.1}s | distance {:.2}",
                    iteration, elapsed, current_distance
                );
            }
        }

        let elapsed = self
            .run_started_at
            .map(|t| t.elapsed())
            .unwrap_or_default();
        self.last_run_elapsed = Some(elapsed);

        let best_distance = self
            .best_solution
            .as_ref()
            .map(|s| self.tour_distance(s))
            .unwrap_or(f64::INFINITY);
        let tour = self.best_solution.clone().unwrap_or_default();

        eprintln!(
            "Finished. Best distance = {:.2}, time = {:.3} s, iterations = {}",
            best_distance,
            elapsed.as_secs_f64(),
            self.last_run_iterations.unwrap_or(0)
        );

        (tour, best_distance)
    }
}
